modules/assim.sequential/R/downscale_function.R

```diff
@@ -1,88 +1,284 @@
+##' @param carbon_pool Character. Carbon pool of interest. Name must match carbon pool name found within file supplied to 'data'.
+##' @param covariates SpatRaster stack. Used as predictors in CNN. Layers within stack should be named.
+##' @param model_type Character. Either "rf" for Random Forest or "cnn" for Convolutional Neural Network.
+##' @details This function will downscale forecast data to unmodeled locations using covariates and site locations
+##'
+##' @description This function uses either Random Forest or Convolutional Neural Network model based on the model_type parameter.
+##'
+##' @return A list containing the training and testing data sets, models, predicted maps for each ensemble member, and predictions for testing data.
+
+SDA_downscale <- function(preprocessed, date, carbon_pool, covariates, model_type) {
+  input_data <- preprocessed$input_data
+  site_coordinates <- preprocessed$site_coordinates
+  carbon_data <- preprocessed$carbon_data
+  
+  # Convert site coordinates to SpatVector
+  site_coordinates <- terra::vect(site_coordinates, geom = c("lon", "lat"), crs = "EPSG:4326")
+  
+  # Extract predictors from covariates raster using site coordinates
+  predictors <- as.data.frame(terra::extract(covariates, site_coordinates, ID = FALSE))
+  
+  # Dynamically get covariate names
+  covariate_names <- names(predictors)
+  
+  # Create a single data frame with all predictors and ensemble data
+  full_data <- cbind(carbon_data, predictors)
+  
+  # Split the observations into training and testing sets
+  set.seed(123)  # for reproducibility
+  sample <- sample(1:nrow(full_data), size = round(0.75 * nrow(full_data)))
+  train_data <- full_data[sample, ]
+  test_data <- full_data[-sample, ]
+  
+  if (model_type == "rf") {
+    # Train a random forest model for each ensemble member using the training data
+    rf_output <- list()
+    for (i in seq_along(carbon_data)) {
+      ensemble_col <- paste0("ensemble", i)
+      formula <- as.formula(paste(ensemble_col, "~", paste(covariate_names, collapse = " + ")))
+      rf_output[[i]] <- randomForest::randomForest(formula,
+                                                   data = train_data,
+                                                   ntree = 1000,
+                                                   na.action = stats::na.omit,
+                                                   keep.forest = TRUE,
+                                                   importance = TRUE)
+    }
+    
+    # Generate predictions (maps) for each ensemble member using the trained models
+    maps <- list()
+    predictions <- list()
+    for (i in seq_along(rf_output)) {
+      maps[[i]] <- terra::predict(covariates, model = rf_output[[i]], na.rm = TRUE)
+      predictions[[i]] <- predict(rf_output[[i]], test_data)
+    }
+    
+    # Organize the results into a single output list
+    downscale_output <- list(
```

Comment:
The object returned by this function should be identical for both options, otherwise it won't be usable.

Reply:
Addressed change in commit [c8c234a](https://github.com/PecanProject/pecan/pull/3321/commits/c8c234a9bff5d484f0c6e62a110765520ce21807)